### PR TITLE
(doc) Add last updated date in docs footer

### LINKED ISSUE #19272: [docs] Add a "Last updated" date to reference in docs (reference guides, recipes, tutorials)
## Description

Related to https://github.com/gatsbyjs/gatsby/pull/19222, an incorrect date of Feb 9, 2011 appears in Google SERP in the [tutorial "Building a Theme"](https://www.gatsbyjs.org/tutorial/building-a-theme/). 

image

### Goal 
To prevent issues like this from happening and to convey transparency on the maintenance of the docs, It may be worth to add a "Last updated" date for docs pages and make it dynamically pull from git.

### PR DESCRIPTION
**Pull request by kraynel**
## Description

Hello! I came across issue #19272 yesterday and found it would be a good way for me to start contributing to your excellent projet.

Here is my take on this, I am not sure the multiple calls to the `simple-git` library is the most efficient way of doing that. What do you think?

~~I am only displaying this date for the `docs` and `tutorials` sections, not for the package as it seems less relevant.~~

<table>
<thead>
<tr>
<th>Desktop</th>
<th>Mobile</th>
</tr>
</thead>
<tbody>
<tr>
<td>
image</td><td>
image</td></tr>
</tbody>
</table>

## Related Issues

Fixes #19272

### CODE REVIEW

**Comment by muescha** [comment]
I would expected to get the time when the file is touched first time in gatsby-source-filesystem and not later a second time - but maybe I am wrong what we can get from filesystem:

https://github.com/gatsbyjs/gatsby/blob/561d33e2e491d3971cb2a404eec9705a5a493602/packages/gatsby-source-filesystem/index.d.ts#L58-L64

**Comment by kraynel** [comment]
Unfortunately [git does not track file modification time](https://git.wiki.kernel.org/index.php/Git_FAQ#Why_isn.27t_Git_preserving_modification_time_on_files.3F), so if I use the timestamp from the `gatsby-source-filesystem`, I got my local modification datetime (or the datetime I cloned the repo).

**Comment by muescha** [comment]
Or Can this be in a Transformer Plugin( with include / exclude Regex options)?

**Comment by kraynel** [comment]
I like the idea of a transformer plugin exposing some git info in addition to the filesystem info. Let's try that.

**Comment by kraynel** [comment]
I just published a new transformer plugin: [gatsby-transformer-gitinfo](https://github.com/kraynel/gatsby-transformer-gitinfo).

This PR is much cleaner, thanks for the feedback!

The last modification date is available as a field info for all `.md` files, as well as the last commit author name and email.

We display that last modification date in the footer of all docs, tutos and official packages.

**Comment by muescha** [comment]
Great!

That s much better!

**Comment by muescha** [comment]
Make it sende to name it 
gitLogLatestDate

Or even something like

```graphql
query {
  allFile {
    edges {
      node {
        fields {
          gitLogLatestDate
          gitLogLatestAuthorName
          gitLogLatestAuthorEmail
        }
        internal {
          type
          mediaType
          description
          owner
        }
      }
    }
  }
}
```


```graphql
query {
  allFile {
    edges {
      node {
        fields {
          git {
            log {
              latest {
                date
                authorName
                authorEmail
              }
            }
          }
        }
        internal {
          type
          mediaType
          description
          owner
        }
      }
    }
  }
}
```


and also include more git parsable information from.git. 

For example the (remote,origin) or something what is there

```graphql
{
  git {
    origin {  // depending on destination name
      url
    }
}

**Comment by muescha** [comment]
then this

```jsx
            <a
              sx={{ variant: `links.muted` }}
              href={`https://github.com/gatsbyjs/gatsby/blob/master/${
                this.props.packagePage ? `packages` : `docs`
              }/${this.props.page ? this.props.page.parent.relativePath : ``}`}
            >
```

can be an:

```jsx
        ( this.props.page.parent.fields.git.origin.url 
			&& 
            <a
              sx={{ variant: `links.muted` }}
              href={this.props.page.parent.fields.git.origin.url}
            >
```

**Comment by kraynel** [comment]
Ok I'll restructure the fields to add more room for improvement.

For the remote info, I think I will not be able to guess the remote http url. I can easily fetch the remote url `https://github.com/gatsbyjs/gatsby.git`, but it can also be checked out using SSH: `git@github.com:gatsbyjs/gatsby.git`.

**Comment by kraynel** [comment]
I am having some difficulties with the git library I am using: https://github.com/steveukx/git-js/issues/302, some log commands are returning an empty history. It seems to do so because we run lots of async spawns at the same time.

I wanted to switch to `nodegit/nodegit`, but they also have issues when fetching a specific file history: https://github.com/nodegit/nodegit/issues/1068. They have a much slower alternative witch requires analysing all commits until the file is found, which is not really usable on a large repo.

**Comment by LekoArts** [comment]
I used this in the past successfully, maybe you want to investigate that as an alternative:
https://gist.github.com/LekoArts/a4ece3918bf452fa1be882d712938a4b

**Comment by kraynel** [comment]
Thanks, Docusaurus is using something similar: https://github.com/facebook/docusaurus/blob/master/packages/docusaurus-1.x/lib/core/utils.js#L41-L96

If i cannot get the `simple-git` version to work I will switch to one of these solutions.

**Comment by kraynel** [comment]
I just noticed something, `createNodeField` works fine when the value is a string but randomly fails when I give it an object. The [doc](https://www.gatsbyjs.org/docs/actions/#createNodeField) says it should be a string. How can I can the `fields.git.log.latest` hierarchy?

I have a working version with that interface, should we keep that for the first version?

```graphql
query {
  allFile {
    edges {
      node {
        fields {
          gitLogLatestDate
          gitLogLatestAuthorName
          gitLogLatestAuthorEmail
        }
        internal {
          type
          mediaType
          description
          owner
        }
      }
    }
  }
}
```

**Comment by LekoArts** [comment]
I think it's fine for now. Let's try to get this thing in :)

**Comment by kraynel** [comment]
Cool! How do you want to proceed? Should I release the 1.0.0 version of the lib then update the `package.json`?

**Comment by kraynel** [comment]
I rebase and pointed the package to a stable version.

**Comment by wardpeet** [comment]
Hey @kraynel do you have any idea how expensive this lookup is? Calling git log each time sounds expensive but maybe it isn't :smile: Just checking if this is the approach we're looking for. Thanks a lot for digging into this!

**Comment by kraynel** [comment]
Hi @wardpeet, I will post some benchmarks to give you an idea of the impact on the build time. Docusaurus is disabling this feature in dev mode, I can probably do the same.

**Comment by LekoArts** [comment]
Hi @kraynel 👋 Do you have time to create those benchmarks? Would be awesome!

**Comment by kraynel** [comment]
Sorry for the delay, trying to run that today.

**Comment by kraynel** [comment]
Hi @LekoArts @wardpeet, did you had time to take look? Can I help you to do something?

**Comment by laurieontech** [comment]
I tested this locally and it looked good from a Learning perspective but will wait for core to assess the performance impacts.

File: www/src/components/markdown-page-footer.js [hunk 1]
```diff
@@ -14,16 +14,38 @@ export default class MarkdownPageFooter extends React.Component {
...
+              }/${this.props.page ? this.props.page.parent.relativePath : ``}`}
+            >
+              <EditIcon sx={{ marginRight: 2 }} />
+              {` `}
+              Edit this page on GitHub
+            </a>
+            {this.props.page &&
+              this.props.page.parent &&
+              this.props.page.parent.fields &&
+              this.props.page.parent.fields.gitLogLatestDate && (
```
Comment (pieh) [inline www/src/components/markdown-page-footer.js]: NIT: could we use optional chaining here?

```diff
-            {this.props.page &&
-              this.props.page.parent &&
-              this.props.page.parent.fields &&
-              this.props.page.parent.fields.gitLogLatestDate && (
+            {this.props.page?.parent?.fields?. gitLogLatestDate && (
```

**Comment by pieh** [comment]
We also need to think about ongoing i18n effort - this method won't work for translations as they are not part of the same repository and are pulled from github directly. This likely won't cause any build issues for them, just that it won't show this information for them.

Adding support for this feature to translated pages is not a blocker (IMO), but we should at least produce action items on how to support that.

File: www/src/components/markdown-page-footer.js (hunk 1 above)
Comment (LekoArts) [reply www/src/components/markdown-page-footer.js]: @kraynel I think after this the PR should be good to go 👍

Comment (kraynel) [reply www/src/components/markdown-page-footer.js]: Updated and rebased!

**Comment by kraynel** [comment]
@wardpeet I updated and rebased to take the latest feedbacks I to account. Is it ok for you?

**Comment by marcysutton** [comment]
This will be so cool to add!! Really looking forward to it. I agree that having action items for internationalization is important, but it should be done in a separate issue so we can keep this moving. Thanks for your persistence on it @kraynel!

**Comment by laurieontech** [comment]
Thanks so much for your work on this! Let's finally get this in.

**Comment by fk** [comment]
🎉 Yay, thanks everybody!

**Pull request merged**